src/databricks/labs/ucx/mixins/cached_workspace_path.py

```diff
@@ -1,139 +1,138 @@
 from __future__ import annotations
 
+import codecs
+import io
 import os
 from collections import OrderedDict
 from collections.abc import Generator
-from io import StringIO, BytesIO
+from io import BytesIO
+from pathlib import PurePosixPath
+from typing import IO, TypeVar
 
 from databricks.sdk import WorkspaceClient
 from databricks.sdk.service.workspace import ObjectInfo
 from databricks.labs.blueprint.paths import WorkspacePath
 
 
-class _CachedIO:
-
-    def __init__(self, content):
-        self._content = content
-        self._index = 0
-
-    def __enter__(self):
-        return self
-
-    def __exit__(self, exc_type, exc_val, exc_tb):
-        return False
-
-    def read(self, *args, **_kwargs):
-        count = -1 if len(args) < 1 or args[0] < 1 else args[0]
-        if count == -1:
-            return self._content
-        start = self._index
-        end = self._index + count
-        if start >= len(self._content):
-            return None
-        self._index = self._index + count
-        return self._content[start:end]
-
-    def __iter__(self):
-        if isinstance(self._content, str):
-            yield from StringIO(self._content)
-            return
-        yield from self._as_string_io().__iter__()
+# lru_cache won't let us invalidate cache entries
+# so we provide our own custom lru_cache
+class _PathLruCache:
 
-    def with_mode(self, mode: str):
-        if 'b' in mode:
-            return self._as_bytes_io()
-        return self._as_string_io()
+    _datas: OrderedDict[PurePosixPath, bytes]
+    """Cached binary data of files, keyed by resolved workspace path, ordered from oldest to newest."""
 
-    def _as_bytes_io(self):
-        if isinstance(self._content, bytes):
-            return self
-        return BytesIO(self._content.encode("utf-8-sig"))
+    _max_entries: int
+    """The maximum number of entries to hold in the cache."""
 
-    def _as_string_io(self):
-        if isinstance(self._content, str):
-            return self
-        return StringIO(self._content.decode("utf-8"))
+    def __init__(self, max_entries: int) -> None:
+        # Ordered from oldest to newest.
+        self._datas = OrderedDict()
+        self._max_entries = max_entries
 
+    @classmethod
+    def _normalize(cls, path: _CachedPath) -> PurePosixPath:
+        # Note: must not return the same instance that was passed in, to avoid circular references (and memory leaks).
+        return PurePosixPath(path.resolve())
```

Comment:
and what happens if `.resolve()` fails?

Reply:
That's a good question. It shouldn't happen: `WorkspacePath.resolve()` only fails when the path is not absolute (because `Workspace.cwd()` doesn't exist). However we now guard against non-absolute paths on the way in.

That said, now that you point it out I'm on the fence about keeping it. Removing it means we might end up double-caching the same thing, but I'm not even sure if that's possible for `WorkspacePath`.

Will remove…

Reply:
… and removed.